### PR TITLE
[Api][Product] Product show 

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

Screenshot 2020-07-22 at 12 42 59

### CODE REVIEW

File: src/Sylius/Component/Core/Formatter/StringInflector.php [hunk 1]
```diff
@@ -20,7 +20,7 @@ public static function nameToCode(string $value): string
         return str_replace([' ', '-', '\''], '_', $value);
     }
 
-    public static function nameToSlug(string $value): string
+    public static function codeToSlug(string $value): string
```
Comment (AdamKasp) [inline src/Sylius/Component/Core/Formatter/StringInflector.php:23]: why did you change it?
is is a BC? If it is you should add a new method :)

Comment (Tomanhez) [reply src/Sylius/Component/Core/Formatter/StringInflector.php:23]: It was not used anywhere, and was not consistent, but you aright, it can be BC break.

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml [hunk 2]
```diff
@@ -47,6 +47,9 @@
         <itemOperations>
             <itemOperation name="get">
                 <attribute name="path">products/{id}</attribute>
+                <attribute name="openapi_context">
+                    <attribute name="summary">As an logged administrator use $code to retrieves a product resource. As an Visitor or logged Customer use $slug to retrieves a product resource.</attribute>
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml:51]: shouldn't we say 'use $code to retrieve'? `retrieves` is used in other tense.  Same goes for $slug

File: src/Sylius/Behat/Client/ResponseChecker.php [hunk 3]
```diff
@@ -48,6 +48,13 @@ public function getValue(Response $response, string $key)
         return $this->getResponseContentValue($response, $key);
     }
 
+    public function getTranslationValue(Response $response, string $key, ?string $locale = 'en_US'): string
```
Comment (GSadee) [inline src/Sylius/Behat/Client/ResponseChecker.php:51]:
```suggestion
    public function getTranslationValue(Response $response, string $key, ?string $localeCode = 'en_US'): string
```

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 4]
```diff
@@ -30,6 +30,9 @@ final class CartContext implements Context
     /** @var ApiClientInterface */
     private $cartsClient;
 
+    /** @var ApiClientInterface */
+    private $productClient;
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:34]:
```suggestion
    private $productsClient;
```
to be consistent with the service above

File: src/Sylius/Behat/Context/Api/Shop/ProductContext.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+     * @Then I should see the product name :name
+     */
+    public function iShouldSeeProductName(string $name): void
+    {
+        Assert::true($this->responseChecker->hasItemWithTranslation(
+            $this->client->getLastResponse(),
+            'en_US',
+            'name',
+            $name))
+        ;
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/ProductContext.php:54]:
```suggestion
        Assert::true(
            $this->responseChecker->hasItemWithTranslation($this->client->getLastResponse(), 'en_US', 'name', $name)
        );
```
?

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml (hunk 2 above)
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml:51]:
```suggestion
                    <attribute name="summary">As a logged administrator use $code to retrieve a product resource. As a Visitor or logged in Customer use $slug to retrieve a product resource.</attribute>
```

File: src/Sylius/Bundle/ApiBundle/spec/DataProvider/ProductItemDataProviderSpec.php [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        $this->beConstructedWith($productRepository, $userContext);
+    }
+
+    function it_supports_only_product(): void
+    {
+        $this->supports(ProductInterface::class, Request::METHOD_GET)->shouldReturn(true);
+        $this->supports(TaxonInterface::class, Request::METHOD_GET)->shouldReturn(false);
+    }
+
+    function it_throws_an_exception_if_context_has_not_channel(): void
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/DataProvider/ProductItemDataProviderSpec.php:40]:
```suggestion
    function it_throws_an_exception_if_context_has_no_channel(): void
```

File: src/Sylius/Bundle/ApiBundle/spec/DataProvider/ProductItemDataProviderSpec.php [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            ->during('getItem', [
+                ProductInterface::class,
+                'ford',
+                Request::METHOD_GET,
+                [ContextKeys::LOCALE_CODE => 'en_US'],
+            ])
+        ;
+    }
+
+    function it_throws_an_exception_if_context_has_not_locale_code(ChannelInterface $channel): void
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/DataProvider/ProductItemDataProviderSpec.php:53]:
```suggestion
    function it_throws_an_exception_if_context_has_no_locale_code(ChannelInterface $channel): void
```

File: src/Sylius/Bundle/ApiBundle/spec/DataProvider/ProductItemDataProviderSpec.php [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            ->during('getItem', [
+                ProductInterface::class,
+                'ford',
+                Request::METHOD_GET,
+                [ContextKeys::CHANNEL => $channel],
+            ])
+        ;
+    }
+
+    function it_provides_product_from_product_id_for_logged_admin(
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/DataProvider/ProductItemDataProviderSpec.php:66]:
```suggestion
    function it_provides_product_by_code_for_a_logged_in_admin_user(
```

File: src/Sylius/Bundle/ApiBundle/spec/DataProvider/ProductItemDataProviderSpec.php [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+                ProductInterface::class,
+                'FORD_FOCUS',
+                Request::METHOD_GET,
+                []
+            )
+            ->shouldReturn($product)
+        ;
+    }
+
+    function it_provides_product_from_product_slug_for_logged_user(
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/DataProvider/ProductItemDataProviderSpec.php:88]:
```suggestion
    function it_provides_product_by_slug_for_a_logged_in_shop_user(
```

File: src/Sylius/Bundle/ApiBundle/spec/DataProvider/ProductItemDataProviderSpec.php [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        $productRepository->findOneByChannelAndSlug($channel, 'en_US','FORD_FOCUS')->willReturn($product);
+
+        $this
+            ->getItem(
+                ProductInterface::class,
+                'FORD_FOCUS',
+                Request::METHOD_GET,
+                [
+                    ContextKeys::CHANNEL => $channel,
+                    ContextKeys::LOCALE_CODE => 'en_US'
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/DataProvider/ProductItemDataProviderSpec.php:107]:
```suggestion
                    ContextKeys::LOCALE_CODE => 'en_US',
```

File: src/Sylius/Bundle/ApiBundle/spec/DataProvider/ProductItemDataProviderSpec.php [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+                ]
+            )
+            ->shouldReturn($product)
+        ;
+    }
+
+    function it_provides_product_from_product_slug_for_null_user(
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/DataProvider/ProductItemDataProviderSpec.php:114]:
```suggestion
    function it_provides_product_by_slug_if_there_is_no_logged_in_user(
```

File: src/Sylius/Behat/Context/Api/Shop/ProductContext.php (hunk 5 above)
Comment (Tomanhez) [reply src/Sylius/Behat/Context/Api/Shop/ProductContext.php]: I chose solution between, because your proposition was too long
Screenshot 2020-07-22 at 12 37 53

**Comment by GSadee** [comment]
Thanks, Tomasz! :1st_place_medal:

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml [hunk 12]
```diff
@@ -47,6 +47,8 @@
         <itemOperations>
             <itemOperation name="get">
                 <attribute name="path">products/{id}</attribute>
+                <attribute name="openapi_context">
+                    <attribute name="summary">As a logged administrator use $code to retrieve a product resource. As a Visitor or logged in Customer use $slug to retrieve a product resource.</attribute>                </attribute>
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml:51]: this closing </attribute> tag has been hidden just like finnish sniper. Can you find it? :D
